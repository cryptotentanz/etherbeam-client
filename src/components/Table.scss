@use '../styles/resources/sizes';
@use '../styles/resources/colors';

.my-table {
  border-collapse: collapse;
  border-style: solid;
  border-width: sizes.$border;
  width: 100%;
  padding: sizes.$spacer;

  th {
    letter-spacing: sizes.$letter-spacing;
  }

  th,
  td {
    border-style: solid;
    border-width: sizes.$border;
    padding: sizes.$spacer-s;
  }

  // &-primary {
  //   background-color: colors.$bg-primary;
  //   border-color: colors.$bd-primary;
  //   color: colors.$fg-light;

  // }

  // &-secondary {
  //   background-color: colors.$bg-secondary;
  //   border-color: colors.$bd-secondary;
  //   color: colors.$fg-light;

  //   th,
  //   td {
  //     border-color: colors.$bd-secondary;
  //   }
  // }

  // &-dark {
  //   background-color: colors.$bg-dark;
  //   border-color: colors.$bd-dark;
  //   color: colors.$fg-dark;

  //   th,
  //   td {
  //     border-color: colors.$bd-dark;
  //   }
  // }

  // &-light {
  //   background-color: colors.$bg-light;
  //   border-color: colors.$bd-light;
  //   color: colors.$fg-light;

  //   th,
  //   td {
  //     border-color: colors.$bd-light;
  //   }
  // }

  // &-positive {
  //   background-color: colors.$bg-positive;
  //   border-color: colors.$bd-positive;
  //   color: colors.$fg-dark;

  //   th,
  //   td {
  //     border-color: colors.$bd-positive;
  //   }
  // }

  // &-negative {
  //   background-color: colors.$bg-negative;
  //   border-color: colors.$bd-negative;
  //   color: colors.$fg-dark;

  //   th,
  //   td {
  //     border-color: colors.$bd-negative;
  //   }
  // }
}
