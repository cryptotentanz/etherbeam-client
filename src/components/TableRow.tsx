import React, { FC, ReactNode } from 'react'

type ColorProp = 'dark' | 'light' | 'primary' | 'secondary' | 'positive' | 'negative'

interface TableRowProps {
  color?: ColorProp
  clickable?: boolean
  action?: () => void
  active?: boolean
  disabled?: boolean
  className?: string
  children?: ReactNode
}

const TableRow: FC<TableRowProps> = ({
  color = 'dark',
  clickable,
  action,
  active,
  disabled,
  className,
  children,
}: TableRowProps) => {
  const onKeyUp = (event: React.KeyboardEvent<HTMLTableRowElement>): void => {
    if (event.key == 'Enter') runAction()
  }

  const runAction = (): void => {
    if (!disabled) action?.()
  }

  const getClassName = (): string => {
    const classNames = [`my-${color}-bg`]
    if (clickable) classNames.push('my-clickable')
    if (disabled) classNames.push('my-disabled-fg')
    if (active) classNames.push('my-active')
    if (className) classNames.push(className)

    return classNames.join(' ')
  }

  return (
    <tr
      onClick={runAction}
      onKeyUp={onKeyUp}
      className={getClassName()}
      tabIndex={clickable && !disabled ? 0 : undefined}
    >
      {children}
    </tr>
  )
}

export default TableRow
