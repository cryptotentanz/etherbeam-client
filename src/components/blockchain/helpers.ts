import { faCoins, faFile, faWallet, IconDefinition } from '@fortawesome/free-solid-svg-icons'
import { BigNumber, ethers } from 'ethers'
import numeral from 'numeral'
import { AddressType, Token } from 'libraries/types'

const ETH_DECIMALS = 18
const MAX_DECIMALS = 13
const UNISWAP_SWAP_URL = 'https://app.uniswap.org/#/swap'

export interface AmountPrintOptions {
  forceDecimals?: boolean
  decimals?: number
}

const printAmount = (amount: string | null, decimals: number, options?: AmountPrintOptions | null): string => {
  const { forceDecimals, decimals: forcedDecimals } = options ?? {}
  const decimalsCount = Math.min(forcedDecimals || decimals, MAX_DECIMALS)
  const decimalFormat = forceDecimals ? '0'.repeat(decimalsCount) : `[${'0'.repeat(decimalsCount)}]`

  return amount ? numeral(amount).format(`0,0.${decimalFormat}`) : '?'
}

export type Unit = 'wei' | 'kwei' | 'mwei' | 'gwei' | 'ether'

export const printEtherAmount = (amount: BigNumber | null, options?: AmountPrintOptions | null, unit?: Unit): string =>
  printAmount(amount ? ethers.utils.formatUnits(amount, unit) : null, ETH_DECIMALS, options)

export const printTokenAmount = (
  amount: BigNumber | null,
  token: Token,
  options?: AmountPrintOptions | null
): string => {
  const { decimals } = token

  return printAmount(amount ? ethers.utils.formatUnits(amount, decimals) : null, decimals, options)
}

export const shortenAddress = (address: string): string => {
  const start = address.substring(0, 6)
  const end = address.substring(address.length - 4)

  return `${start}...${end}`
}

export const getAddressTypeIcon = (type: AddressType): IconDefinition | null => {
  switch (type) {
    case AddressType.Wallet:
      return faWallet
    case AddressType.Contract:
      return faFile
    case AddressType.Token:
      return faCoins
    default:
      return null
  }
}

export const getUniswapTradeUrl = (inputHash: string | null, outputHash: string | null): string => {
  return `${UNISWAP_SWAP_URL}?${inputHash ? `inputCurrency=${inputHash}&` : ''}${
    outputHash ? `outputCurrency=${outputHash}` : ''
  }`
}
