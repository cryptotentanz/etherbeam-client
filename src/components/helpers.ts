import { UserType } from 'libraries/types'
import { faUser, faUserShield, IconDefinition } from '@fortawesome/free-solid-svg-icons'

export const getUserTypeIcon = (type: UserType): IconDefinition => {
  switch (type) {
    case UserType.Administrator:
      return faUserShield
    default:
      return faUser
  }
}
