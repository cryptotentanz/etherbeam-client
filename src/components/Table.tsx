import React, { FC, ReactNode } from 'react'
import './Table.scss'

type ColorProp = 'dark' | 'light' | 'primary' | 'secondary' | 'positive' | 'negative'

interface TableProps {
  color?: ColorProp
  children?: ReactNode
}

const Table: FC<TableProps> = ({ color = 'dark', children }: TableProps) => {
  return <table className={`my-table my-${color}-bg`}>{children}</table>
}

export default Table
