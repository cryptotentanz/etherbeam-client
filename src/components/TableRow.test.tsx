import React from 'react'
import { mount, ReactWrapper } from 'enzyme'
import { act } from 'react-dom/test-utils'
import { clearMocks } from 'tests/helpers'
import Table from './Table'
import TableRow from './TableRow'

describe('<TableRow />', () => {
  const action = jest.fn()
  const content = <td></td>
  const className = 'my-class'

  let component: ReactWrapper

  describe('Default', () => {
    beforeAll(() => {
      component = mount(
        <Table>
          <tbody>
            <TableRow>{content}</TableRow>
          </tbody>
        </Table>
      )
    })

    it('should render', () => {
      const tr = component.find('tr')
      expect(tr.prop('tabIndex')).toBeUndefined()
      expect(tr.prop('children')).toBe(content)
      expect(tr.hasClass('my-dark-bg')).toBeTruthy()
    })

    it('should match snapshot', () => expect(component.render()).toMatchSnapshot())
  })

  describe('Color', () => {
    beforeAll(() => {
      component = mount(
        <Table>
          <tbody>
            <TableRow color="primary">{content}</TableRow>
          </tbody>
        </Table>
      )
    })

    it('should render', () => expect(component.find('tr').hasClass('my-primary-bg')).toBeTruthy())
  })

  describe('Clickable', () => {
    beforeAll(() => {
      component = mount(
        <Table>
          <tbody>
            <TableRow clickable action={action}>
              {content}
            </TableRow>
          </tbody>
        </Table>
      )
    })

    it('should render', () => {
      const tr = component.find('tr')
      expect(tr.prop('tabIndex')).toBe(0)
      expect(tr.hasClass('my-clickable')).toBeTruthy()
    })
  })

  describe('Active', () => {
    beforeAll(() => {
      component = mount(
        <Table>
          <tbody>
            <TableRow active>{content}</TableRow>
          </tbody>
        </Table>
      )
    })

    it('should render', () => expect(component.find('tr').hasClass('my-active')).toBeTruthy())
  })

  describe('Disabled', () => {
    beforeAll(() => {
      component = mount(
        <Table>
          <tbody>
            <TableRow clickable action={action} disabled>
              {content}
            </TableRow>
          </tbody>
        </Table>
      )
    })

    it('should render', () => {
      const tr = component.find('tr')
      expect(tr.prop('tabIndex')).toBeUndefined()
      expect(tr.hasClass('my-disabled-fg')).toBeTruthy()
    })
  })

  describe('Class', () => {
    beforeAll(() => {
      component = mount(
        <Table>
          <tbody>
            <TableRow className={className}>{content}</TableRow>
          </tbody>
        </Table>
      )
    })

    it('should render', () => expect(component.find('tr').hasClass('my-class')).toBeTruthy())
  })

  describe('On click', () => {
    beforeAll(() => {
      clearMocks()

      component = mount(
        <Table>
          <tbody>
            <TableRow clickable action={action}>
              {content}
            </TableRow>
          </tbody>
        </Table>
      )

      act(() => {
        component.find('tr').simulate('click')
        component.update()
      })
    })

    it('should call action', () => expect(action).toBeCalledTimes(1))
  })

  describe('On key up (enter)', () => {
    beforeAll(() => {
      clearMocks()
      component = mount(
        <Table>
          <tbody>
            <TableRow clickable action={action}>
              {content}
            </TableRow>
          </tbody>
        </Table>
      )

      act(() => {
        component.find('tr').simulate('keyUp', { key: 'Enter' })
        component.update()
      })
    })

    it('should call action', () => expect(action).toBeCalledTimes(1))
  })

  describe('On key up (other)', () => {
    beforeAll(() => {
      clearMocks()
      component = mount(
        <Table>
          <tbody>
            <TableRow clickable action={action}>
              {content}
            </TableRow>
          </tbody>
        </Table>
      )

      act(() => {
        component.find('tr').simulate('keyUp', { key: 'a' })
        component.update()
      })
    })

    it('should not call action', () => expect(action).toBeCalledTimes(0))
  })
})
