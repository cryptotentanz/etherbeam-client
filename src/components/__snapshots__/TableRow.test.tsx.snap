// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TableRow /> Default should match snapshot 1`] = `
<table
  class="my-table my-dark-bg"
>
  <tbody>
    <tr
      class="my-dark-bg"
    >
      <td />
    </tr>
  </tbody>
</table>
`;
